Mathlib/Analysis/SpecialFunctions/Log/NegMulLog.lean

```diff
@@ -52,18 +52,128 @@ lemma hasDerivAt_mul_log {x : ℝ} (hx : x ≠ 0) : HasDerivAt (fun x ↦ x * lo
+
+-- lemma isLittleO_nhds_zero_id_deriv_mul_log : id =o[𝓝[>] 0] (deriv (fun x ↦ x * log x)) := by
+--   sorry
+
+/- helper lemma for `not_eventually_bounded_zero_mul_log` -/
+private lemma NegMulLog.one_lt_log_sub_const_of_lt_exp_sub
+    (x D : ℝ) (hx : 0 < x) (h : x < rexp (D - 2)) :
+    1 < |x.log - D| := by
+  by_cases abs_eq : x.log - D < 0
+  · rw [abs_of_neg]
+    have : x.log - D < -2 := sub_left_lt_of_lt_add ((log_lt_iff_lt_exp hx).mpr h)
+    linarith
+    exact abs_eq
+  · have h := log_lt_log hx h
+    simp only [log_exp] at h
+    rw [abs_of_pos (by linarith)]
+    linarith
+
+lemma not_eventually_bounded_derivative_zero_mul_log (D : ℝ) :
+    ¬ ∀ᶠ (x : ℝ) in 𝓝 0, |x * x.log - x * D| ≤ |x| := by
+  simp [eventually_nhdsWithin_iff, Metric.eventually_nhds_iff]
+  intro x hx
+  exists min (2⁻¹ * x) (exp (D - 2))
+  have xhalf_gt0 : 0 < 2⁻¹ * x := by norm_num [hx]
+  constructor
+  · rw [abs_of_nonneg (le_min (le_of_lt xhalf_gt0) (exp_nonneg (D - 2)))]
+    apply min_lt_of_left_lt
+    norm_num
+    linarith
+  · have exists_pos : 0 < |min (2⁻¹ * x) (rexp (D - 2))| := by
+      simp only [one_div, abs_pos]
+      exact ne_of_gt (lt_min xhalf_gt0 (exp_pos (D - 2)))
+    simp only [← mul_sub, one_div, abs_mul]
+    apply (lt_mul_iff_one_lt_right exists_pos).mpr
+    by_cases min_l : (2⁻¹ * x) < (rexp (D - 2))
+    · rw [min_eq_left_of_lt min_l]
+      exact NegMulLog.one_lt_log_sub_const_of_lt_exp_sub (2⁻¹ * x) D xhalf_gt0 min_l
+    · simp only [min_eq_right (le_of_not_lt min_l), log_exp, sub_sub_cancel_left, abs_neg,
+        Nat.abs_ofNat, Nat.one_lt_ofNat]
+
+-- TODO inline or refactor or find home
+/- In the hypothesis, D stands for a hypothetical derivative `deriv f x0`. -/
+lemma not_DifferentiableAt_of_not_eventuallly_bounded_derivative (f : ℝ → ℝ) (x0 : ℝ)
+    (hf : ∀ D, ¬ ∀ᶠ (x : ℝ) in 𝓝 0, |f (x0 + x) - f x0 - x * D| ≤ |x|) :
+    ¬ DifferentiableAt ℝ f x0 := by
+  intro h
+  have := hasDerivAt_iff_isLittleO_nhds_zero.mp (DifferentiableAt.hasDerivAt h)
+  simp only [zero_add, log_zero, mul_zero, sub_zero, smul_eq_mul] at this
+  have := Asymptotics.IsLittleO.bound this
+  simp only [norm_mul, norm_eq_abs] at this
+  have := @this 1 (by norm_num)
+  simp only [one_mul] at this
+  have := hf (deriv f x0)
+  contradiction
+
+lemma not_DifferentiableAt_log_mul_zero :
+    ¬ DifferentiableAt ℝ (fun x ↦ x * log x) 0 := by
+  apply not_DifferentiableAt_of_not_eventuallly_bounded_derivative (fun x ↦ x * log x) 0
+  simp only [zero_add, log_zero, mul_zero, sub_zero]
+  exact not_eventually_bounded_derivative_zero_mul_log
+
+/-- Not differentiable, hence `deriv` has junk value zero. -/
+lemma deriv_mul_log_zero : deriv (fun x ↦ x * log x) 0 = 0 :=
+  deriv_zero_of_not_differentiableAt not_DifferentiableAt_log_mul_zero
+
+section NotContinuousAtOfTendstoNhdsWithin
+-- TODO put elsewhere, generalize, maybe add other nhdsWithin variants?
+
+open Filter
+
+lemma not_continuousAt_of_tendsto_nhdsWithin_Ioi_atTop {f : ℝ → ℝ} {x : ℝ}
```

Comment:
@dupuisf  said (in https://github.com/leanprover-community/mathlib4/pull/9734#discussion_r1669502767)
> I think these could be generalized to something like
> 
> ```
> lemma not_continuousAt_of_tendsto {f : ℝ → ℝ} {l₁ l₂ : Filter ℝ} (x : ℝ)
>     (hf : Tendsto f l₁ l₂) [l₁.NeBot] (hl₁ : l₁ ≤ 𝓝 x) (hl₂ : l₂ ⊓ 𝓝 (f x) = ⊥) :
>     ¬ ContinuousAt f x := by sorry
> ```
> 
> (but properly generalized to more than just real functions). Or they could just be inlined since the proofs are quite short anyway. I think having the lemma would be nice, since it's a bit more discoverable than having to resort to lemmas about tendsto.

Reply:
Assuming I could prove that, not sure how to use this not_continuousAt_of_tendsto. For example, to specialize to my case, automation doesn't seem to do it and the best I could come up with is
```lean
lemma not_continuousAt_of_tendsto_nhdsWithin_Ioi_atTop {f : ℝ → ℝ} {x : ℝ}
    (hf : Tendsto f (𝓝[>] x) atTop) :
    ¬ ContinuousAt f x := by
  apply not_continuousAt_of_tendsto x hf nhdsWithin_le_nhds
  apply inf_eq_bot_iff.mpr
  use Set.Ioi (f x + 2)
  constructor
  · exact Ioi_mem_atTop (f x + 2)
  · use Set.Ioo (f x - 1) (f x + 1)
    constructor
    · exact Ioo_mem_nhds (by linarith) (by linarith)
    · rw [Set.Ioi_inter_Ioo]
      simp only [max_lt_iff, add_lt_add_iff_left, Nat.not_ofNat_lt_one, false_and,
        not_false_eq_true, Set.Ioo_eq_empty]
``
Ideally, one would need more lemmas (or something else?) such that exact? or aesop can prove this.

Reply:
`inf_nhds_atTop` should kill off that last one.

Reply:
Any way we can get automation to prove this? As it stands currently, having  `not_continuousAt_of_tendsto` in Mathlib would most likely not have helped me prove `not_continuousAt_of_tendsto_nhdsWithin_Ioi_atTop`.

Reply:
Probably the easiest way to get automation to work is to have `inf_atTop_nhds` as well, since `inf_comm` would be a bad simp lemma.

Reply:
Sounds good. Can I add `inf_atTop_nhds` in this PR or does it need another separate one?

Reply:
I'm confused about the hypothesis above: `hl₁ : l₁ ≤ 𝓝 x`. I can prove it with the comparison reversed
```lean
lemma not_continuousAt_of_tendsto {f : ℝ → ℝ} {l₁ l₂ : Filter ℝ} (x : ℝ)
    (hf : Tendsto f l₁ l₂) [l₁.NeBot] (hl₁ : 𝓝 x ≤ l₁) (hl₂ : l₂ ⊓ 𝓝 (f x) = ⊥) :
    ¬ ContinuousAt f x := fun cont ↦ 
  (cont.tendsto.not_tendsto (disjoint_iff.mpr hl₂).symm) (hf.mono_left hl₁)
```
which seems to suggest that either the hypothesis is not needed or the lemma you stated is false?

Reply:
Both are true and I proved the other direction:

```lean
lemma not_continuousAt_of_tendsto {f : ℝ → ℝ} {l₁ l₂ : Filter ℝ} (x : ℝ)
    (hf : Tendsto f l₁ l₂) [l₁.NeBot] (hl₁ : l₁ ≤ 𝓝 x) (hl : l₂ ⊓ 𝓝 (f x) = ⊥) :
    ¬ ContinuousAt f x := fun cont ↦
  (((ContinuousAt.tendsto_iff.mp cont).mono_left hl₁).not_tendsto (disjoint_iff.mpr hl₂).symm) hf
```

@dupuisf how to proceed with this? Should I add these lemmas to the PR? Or will you make a completely different API for proving non-differentiable and these lemmas are unnecessary?

I guess `hl₂` could be replaced by `Disjoint`, though I'm not sure if that's helpful or not.  Could the two lemmas (both comparisons) be combined into one in some nice way?

Reply:
Hmm, I think the right way to combine `l₁ ≤ 𝓝 x` and `𝓝 x ≤ l₁` is `ClusterPt x l₁`, and likewise `hl₂` would be `hl₂ : ¬ ClusterPt (f x) l₂`.

Reply:
Thanks for the suggestion. On first glance, it doesn't look like `ClusterPt` has a lot of lemmas about it that can be used here (e.g., that allow an easy proof of `not_continuousAt_of_tendsto_nhdsWithin_Ioi_atTop` using a version of `not_continuousAt_of_tendsto` that uses `ClusterPt`). Maybe I'll stick with the current version so-far.

Reply:
I moved a version of `not_continuousAt_of_tendsto` that uses `Disjoint` but not `ClusterPt` to `Mathlib/Topology/Basic.lean`.